[doc] Dashboard documentation

rkooo567 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

Dashboard documentation.

NOTE:
- Images are hosted here. It is to reduce the size of repository because we don't want to have long `git clone` time. https://github.com/ray-project/Images/tree/master/docs/dashboard


## Related issue number


## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22345/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22351/
Test FAILed.

richardliaw commented:
I think you should merge Intro + Getting Started + Concepts into one page; there's not enough information to warrant separate pages for now (and it also seems to clutter the sidebar).

robertnishihara commented on doc/source/dashboard/dashboard-usage.rst:
```diff
@@ -0,0 +1,199 @@
+.. image:: https://raw.githubusercontent.com/ray-project/images/master/docs/dashboard/dashboard-profiling-buttons.png
+    :align: center
+
+Clicking the profling button on the dashboard launches py-spy that times your python program. The timing information will be visualized as flamegraph in a new browser tab.
```
```suggestion
Clicking one of the profiling buttons on the dashboard launches py-spy, which will profile your actor process for the given duration. Once the profiling has been done, you can click the "profiling result" button to visualize the profiling information as a flamegraph.
```

robertnishihara commented on doc/source/dashboard/dashboard-usage.rst:
```diff
@@ -0,0 +1,199 @@
+
+Clicking the profling button on the dashboard launches py-spy that times your python program. The timing information will be visualized as flamegraph in a new browser tab.
+
+Checkout the example Learning to play Pong on ray documentation: `Pong Example <https://ray.readthedocs.io/en/latest/auto_examples/plot_pong_example.html>`_
```
Probably not necessary to link to this example.

robertnishihara commented on doc/source/dashboard/dashboard-usage.rst:
```diff
@@ -0,0 +1,199 @@
+
+Checkout the example Learning to play Pong on ray documentation: `Pong Example <https://ray.readthedocs.io/en/latest/auto_examples/plot_pong_example.html>`_
+
+Click profiling, and click Profiling result when it is ready. Note that there could be multiple threads in the process and some are ray internal threads and the timing information may not be so interesting. Click the left and right arrow on the middle top to see profiling results on different threads.
```
I don't think this paragraph is needed.

robertnishihara commented on doc/source/dashboard/dashboard-usage.rst:
```diff
@@ -0,0 +1,199 @@
+Checkout the example Learning to play Pong on ray documentation: `Pong Example <https://ray.readthedocs.io/en/latest/auto_examples/plot_pong_example.html>`_
+
+Click profiling, and click Profiling result when it is ready. Note that there could be multiple threads in the process and some are ray internal threads and the timing information may not be so interesting. Click the left and right arrow on the middle top to see profiling results on different threads.
+Now you can intuitively see where could be the computation bottleneck. 
```
```suggestion
This visualization can help reveal computational bottlenecks.
```

robertnishihara commented on doc/source/dashboard/ray-dashboard.rst:
```diff
@@ -0,0 +1,18 @@
+Ray Dashboard
```
Why not just have a single documentation page on the dashboard that combines all of the information from the above three pages. Do you think it would be too long?

richardliaw replied:
+1, I think it is better to keep it in one page...

rkooo567 replied:
Three reasons.
1. I personally prefer short pages when I read documentation.
2. I expected each pages will be much longer, but it turns out it was much shorter. 
3. I thought dashboard document could be eventually very long as we will provide lots of different components.

But given the current length, it totally makes sense to put everything within one page. I will merge them.

rkooo567 commented:
@richardliaw I agree! Would you think everything within one page makes more sense than two pages; Intro + Getting Started + Concepts and Usage?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22505/
Test FAILed.

rkooo567 commented:
@robertnishihara I addressed most of comments you left. For things that were not resolved, you can think they were not addressed yet. 

Here are additional improvement I will have.
- Explain Term definition of local memory inspect section.
- Improve Explanation of node memory inspection section.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22802/
Test PASSed.

rkooo567 commented:
@robertnishihara Please take a look at it! I addressed all of your comments and improved memory usage sections. Let me know if you want some more improvement.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22869/
Test PASSed.

richardliaw commented on doc/source/index.rst:
```diff
@@ -283,6 +283,12 @@ Getting Involved
+   :maxdepth: -1
+   :caption: Ray Dashboard
+
+   ray-dashboard.rst
```
can you move this to a page under Ray Core (specifically right under `configure.rst`)

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+Ray Dashboard
+==============
```
```suggestion
=============
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+---------------
+  
+Access Dashboard
+~~~~~~~~~~~~~~~~
```
```suggestion
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+
+You can see that the dashboard shows the parent/child relationship as expected. 
+
+.. image:: https://raw.githubusercontent.com/ray-project/images/master/docs/dashboard/dashboard-tree-actors.png
```
Is it ok to actually move these images into the Ray repo? We currently do that with all of our other images.

robertnishihara replied:
I suggested putting them in the other repo because I was concerned about their size. I imagine we'll want to add a bunch more images as time goes on, and my concern is that cloning the Ray repository will become too slow.

robertnishihara replied:
I imagine this is an issue that a lot of projects run into, so there is probably some standard solution that I'm not aware of.

richardliaw replied:
Seems like the standard solution is to not have that many images :) (just checked OpenAI and pytorch repos, everyone bundles images together, but they don't have that many).

richardliaw replied:
Anyways, I'm not opposed to the idea, but I would suggest a 1 page proposal/memo for this practice to be shared with the Ray team (or else no one knows about it).

rkooo567 replied:
Will create a proposal and write it to slack open-source channel.

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+Ray Dashboard
+==============
+Welcome to the Ray dashboard page! 
```
Let's get rid of this sentence. It feels inconsistent with the tone of the rest of the documentation.

robertnishihara commented:
Was the file `doc/_build/.gitkeep` removed accidentally or intentionally?

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+Welcome to the Ray dashboard page! 
+
+Ray's built-in dashboard provides metrics, charts, and other features that help Ray users to understand Ray clusters and libraries.
+This includes cluster node level metrics, actor level metrics, configurations, profiling & killing actor buttons, and Ray library metrics such as Tune metrics. 
```
```suggestion
This includes node level metrics, actor level metrics, cluster configuration information, tools for profiling and killing actors, and metrics for Ray libraries such as Tune.
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+
+Ray's built-in dashboard provides metrics, charts, and other features that help Ray users to understand Ray clusters and libraries.
+This includes cluster node level metrics, actor level metrics, configurations, profiling & killing actor buttons, and Ray library metrics such as Tune metrics. 
+Check out the documentation to understand a rich set of features we offer.
```
I'd remove this sentence.

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+  
+Access Dashboard
+~~~~~~~~~~~~~~~~
+If you start Ray on your laptop, dashboard is running by default, and you can access it through its URL. 
```
```suggestion
If you start Ray on your laptop, the dashboard runs by default, and you can access it through its URL. 
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+Access Dashboard
+~~~~~~~~~~~~~~~~
+If you start Ray on your laptop, dashboard is running by default, and you can access it through its URL. 
+By default the URL is `localhost:8265`. Note that the port will be changed if the default port is not available.
```
```suggestion
By default the URL is **localhost:8265**. Note that the port will be changed if the default port is not available.
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+
+.. code-block:: text
+
+  INFO resource_spec.py:212 -- Starting Ray with 4.2 GiB memory available for workers and up to 2.12 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
```
Remove this line

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+  INFO resource_spec.py:212 -- Starting Ray with 4.2 GiB memory available for workers and up to 2.12 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
+  INFO services.py:1093 -- View the Ray dashboard at localhost:8265
+
+Dashboard is also available when using autoscaler. Check out `Monitoring Autoscaler Status <autoscaling.html#monitoring-cluster-status>`_ for more details.
```
```suggestion
The dashboard is also available when using the autoscaler. Read about how to `use the dashboard with the autoscaler <autoscaling.html#monitoring-cluster-status>`_.
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+Example Usage
+-------------
+
+To understand each components, look at `Dashboard Components <ray-dashboard.html#components>`_
```
Remove this

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~
+If creating an actor requires resources (e.g., CPUs, GPUs, or other custom resources) that are not currently available, the actor cannot be created until those resources are added to the cluster or become available.
+This can cause an application to hang. To alert you to this issue, infeasible tasks are shown in red in the dashboard, and pending tasks are shown in yellow.
+You can find this information from the `Logical View` tab. 
```
```suggestion
You can find this information from the *Logical View* tab in the dashboard.
```

single back ticks don't do anything in RST I believe

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+This can cause an application to hang. To alert you to this issue, infeasible tasks are shown in red in the dashboard, and pending tasks are shown in yellow.
+You can find this information from the `Logical View` tab. 
+
+- **Infeasible Actor Creation** happens when an actor requires more resources than a Ray cluster can provide.
```
I'd remove this

rkooo567 replied:
Why would you want to remove this definition? Do you think these two are clear enough to distinguish without extra explanation?

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+You can find this information from the `Logical View` tab. 
+
+- **Infeasible Actor Creation** happens when an actor requires more resources than a Ray cluster can provide.
+- **Pending Actor Creation** happens when there is no available resource for this actor because it is already taken by other tasks / actors.
```
I'd remove this

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+.. image:: https://raw.githubusercontent.com/ray-project/Images/master/docs/dashboard/dashboard-pending-infeasible-actors.png
+    :align: center
+
+This cluster has 2 GPUs. Actor1 requires 1 GPU to be created. Actor2 creates 4 Actor1. As a result of calling `Actor2.remote()`, 2 Actor1 creation tasks became pending. 
```
```suggestion
This cluster has two GPUs, and so it only has room to create two copies of ``Actor1``.
```

rkooo567 replied:
I will use this. 

This cluster has two GPUs, and so it only has room to create two copies of ``Actor1``. As a result, the rest of Actor1 will be pending.

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+
+This cluster has 2 GPUs. Actor1 requires 1 GPU to be created. Actor2 creates 4 Actor1. As a result of calling `Actor2.remote()`, 2 Actor1 creation tasks became pending. 
+
+You can also see it is infeasible to create Actor3 because it requires 4 GPUs which is bigger than the total gpus available in this cluster (2 GPUs). 
```
remove this

rkooo567 replied:
Same question as above! : "Why would you want to remove this definition? Do you think these two are clear enough to distinguish without extra explanation?"

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+~~~~~~~~~~~~~~~~~~~~
+The logical view displays the following information about memory usage
+
+- **Number of object IDs in scope**: number of local and remote objectIDs which actors can access.
```
Is it the case that if any of these three numbers are increasing without bound, then there is a problem?

I still don't really understand these three definitions, so I'd suggest remove this section and just saying something like:
>**NumObjectIdsInScope**, **NumLocalObjects**, and **UsedLocalObjectMemory** refer to internal Ray concepts related to memory usage. If any of these numbers is increasing without bound, then there may be a memory leak in the application.

rkooo567 replied:
If I am a developer who uses a dashboard, I'd prefer to know exactly what each metrics means. If you think this definition is confusing, and there is not way to improve, I think we should remove it from dashboard. Otherwise we should explain in the best details so users can use metrics for debugging.

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+- **Pending Actor Creation** happens when there is no available resource for this actor because it is already taken by other tasks / actors.
+
+.. code-block:: python
+  
```
```suggestion
  import ray
  
```

so that users can actually run this

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+    :align: center
+
+Components
+-----------
```
```suggestion
----------
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+The machine view provides node and process level information. You can see resource usage statistics for each node and each worker process.
+
+Logical View (Experimental)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+  
+  ray.init(num_gpus=2)
+
+  @ray.remote(num_gpus=1)
```
For this example, can you get rid of the nesting? It just adds complexity to the example.

rkooo567 replied:
Good point. Will remove this.

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+Logical View (Experimental)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The logical view renders actor information in a tree format. To illustrate this, in the code block below, the ``Parent`` actor creates two ``Child`` actors and each ``Child`` actor creates one ``GrandChild`` actor.
+This relationship will be visible at a dashboard logical view page.
```
```suggestion
This relationship is visible in the dashboard *Logical View* tab.
```

robertnishihara commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,157 @@
+.. image:: https://raw.githubusercontent.com/ray-project/images/master/docs/dashboard/dashboard-tree-actors.png
+    :align: center
+
+For more information about what logical view provides, checkout `Dashboard Usage <ray-dashboard.html#example-usage>`_
```
Remove this

rkooo567 commented:
@robertnishihara I think it is deleted accidentally. It seems to be deleted when I run `make html`. I created a file again.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22907/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23447/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23449/
Test FAILed.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+Start time of each trial.
+
+Hyperparameters
+###############
```
Can you do *bold* instead of a new header? The new header takes up a lot of vertical space: 
```rst
**Trial ID**: Trial IDs for hyperparameter tuning.
**Job ID**: Job IDs for hyperparameter tuning.
**Status**: Status of each trial
**Start Time**: Start time of each trial.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+Yellow colored actors. It happens when there is no available resource for this actor because it is already taken by other tasks / actors.
+
+Actor Hierarchy
+###############
```
Can you use `**bold**`, like:

```rst
**profile for**: Button to run profiling. We currently support 10s, 30s and 60s profiling. It requires passwordless ``sudo``.
**Infeasible Actor Creation**: Red colored actors. It happens when an actor requires more resources than a Ray cluster can provide.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+Errors
+######
+Error messages at each node and worker. You can see error messages by clicking it.
```
Can you **bold** instead of using a header?

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+Let's see how to use this information.
+
+if NumObjectIdsInScope, NumLocalObjects, or UsedLocalObjectMemory keeps growing without bound, it can lead to OOM errors or eviction of objectIDs that your program still wants to use. 
```
Maybe remove L134?

richardliaw replied:
For this entire Inspect Memory Usage section, I would just say:

```
You can detect local memory anomalies through the Logical View tab. if NumObjectIdsInScope, NumLocalObjects, or UsedLocalObjectMemory keeps growing without bound, it can lead to OOM errors or eviction of objectIDs that your program still wants to use. 

.. image..
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+.. image:: https://raw.githubusercontent.com/ray-project/Images/master/docs/dashboard/dashboard-component-view.png
+    :align: center
+
```
Can you add a new H2 header here:

```
Views
-----

```

rkooo567 replied:
Added!

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+- Understand worker and machine behaviors by providing resource usages and status of them.
+- Visualize the actor relationship and stats.
+- Communicate to your Ray cluster. For example, you can kill your actors, profiling application, or many others.
```
```suggestion
- Kill actors and profile your Ray job.
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+- `Logical View <ray-dashboard.html#id2>`_ to see actor stats and hierarchy.
+- `Ray Config <ray-dashboard.html#id3>`_ to see autoscaler configuration. 
+- `Tune <ray-dashboard.html#id4>`_ to see running Tune job status.
+
```
this is a bit redundant with L5-12; I think you can remove this?

rkooo567 replied:
Agree. I removed them.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+Logical View
+~~~~~~~~~~~~
+Go to the Logical View tab when you want to see actor statistics. You can see 
```
The Logical View shows you:

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+~~~~~~~~~~~~
+Check the Machine View tab when you want to see resource usage statistics for each node and each worker process.
+
+It provides
```
The Machine View shows you:

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+Ray Config
+~~~~~~~~~~
+You can see the autoscaler configuration at the Ray Config tab. Look at the picture to see what it provides.
```
The Ray Config tab shows you the current autoscaler configuration.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+Tune
+~~~~
+Go to the Tune tab to see running Tune jobs and see their statistics in details. It offers
```
The Tune tab shows you:

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+If creating an actor requires resources (e.g., CPUs, GPUs, or other custom resources) that are not currently available, the actor cannot be created until those resources are added to the cluster or become available.
+This can cause an application to hang. To alert you to this issue, infeasible tasks are shown in red in the dashboard, and pending tasks are shown in yellow.
+You can find this information from the *Logical View* tab in the dashboard.
```
```suggestion
```

this is a repeat from the first sentence.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+If creating an actor requires resources (e.g., CPUs, GPUs, or other custom resources) that are not currently available, the actor cannot be created until those resources are added to the cluster or become available.
+This can cause an application to hang. To alert you to this issue, infeasible tasks are shown in red in the dashboard, and pending tasks are shown in yellow.
+You can find this information from the *Logical View* tab in the dashboard.
+
```
Below is an example:

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+NumObjectIdsInScope
+###################
+Number of object IDs in scope for this actor. 
+In scope means this object IDs are not supposed to be evicted. For example, if the objectID is locally referenced, it is in scope because this objectID should not be garbage collected to ensure correctness of Ray applications.
```
```suggestion
Number of object IDs in scope for this actor. Object IDs in scope will not be evicted.
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,357 @@
+
+Infeasible Actor Creation
+#########################
+Red colored actors. It happens when an actor requires more resources than a Ray cluster can provide.
```
```suggestion
Actor creation is infeasible when an actor requires more resources than a Ray cluster can provide. This is depicted as a red colored actor.
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+Tune
+~~~~
+Go to the Tune to see running Tune jobs and their statistics.
+
```
reading this now feels awk; I think you can remove it without any loss of info.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+Logical View
+~~~~~~~~~~~~
+Go to the Logical View when you want to see actor statistics. 
+
```
reading this now feels awk; I think you can remove it without any loss of info.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+Machine View
+~~~~~~~~~~~~
+Go to the Machine View when you want to see resource usage statistics for each node and each worker process.
+
```
reading this now feels awk; I think you can remove it without any loss of info.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+Machine View
+~~~~~~~~~~~~
+
+**Machine/Worker Hierarchy**
```
```suggestion
**Machine/Worker Hierarchy**: The dashboard visualizes hierarchical relationship of workers (processes) and machines (nodes). Each host consists of many workers, and you can see them by clicking a + button.
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+
+Ray consists of nodes and workers. Nodes typicall mean machines and workers mean processes.
+The dashboard visualizes hierarchical relationship of workers and machines. Each host consists of many workers, and you can see them by clicking a + button.
+
```
```suggestion

```

richardliaw replied:
(moved above)

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+
+**Actor Hierarchy**
+
+The logical view renders actor information in a tree format. To illustrate this, in the code block below, the ``Parent`` actor creates two ``Child`` actors and each ``Child`` actor creates one ``GrandChild`` actor.
```
```suggestion
**Actor Hierarchy**: The logical view renders actor relationships in a tree format. 

To illustrate this, in the code block below, the ``Parent`` actor creates two ``Child`` actors and each ``Child`` actor creates one ``GrandChild`` actor.
```

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,294 @@
+
+Ray Config
+~~~~~~~~~~~~
+Configuration defined at ``cluster.yaml`` for the autoscaler mode. Check `Cluster.yaml reference <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-full.yaml>`_ to learn it in more details.
```
```suggestion
Configuration defined at ``cluster.yaml`` for the autoscaler mode. See `Cluster.yaml reference <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-full.yaml>`_ for more details.
```

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23462/
Test PASSed.

richardliaw commented on doc/source/ray-dashboard.rst:
```diff
@@ -0,0 +1,284 @@
+**Host**: If it is a node, it shows host information. If it is a worker, it shows a pid.
+
+**Workers**: If it is a node, it shows a number of workers and virtual cores. Note that number of workers can exceed number of cores.
+**Uptime**: Uptime of each worker and process.
```
```suggestion

**Uptime**: Uptime of each worker and process.
```

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23470/
Test PASSed.
